Bug: Add validation in schema Fields in case of name conflicts with mutation

My Fields is as follows:

```
func (Phone) Fields() []ent.Field {
	return []ent.Field{
		field.Int("user_id").Positive(),
		field.String("op"),
		field.String("number").MaxLen(11),
	}
}
```

(I have changed Phone Fields some times, e.g. add user_id field) 

after i run "go generete ./ent ", there is a file named "mutation.go", I get the code as follows:

```
type PhoneMutation struct {
	config
	op            Op
	typ           string
	id            *int
	user_id       *int
	adduser_id    *int
	op            *string
	number        *string
	clearedFields map[string]struct{}
	done          bool
	oldValue      func(context.Context) (*Phone, error)
}
```

you can see it, in the PhoneMutation, i have two fields named "op",  when i run my app, the error "**ent/mutation.go:37:2: duplicate field op**" will be raised, 

so i think if someone add field named like "typ","done"... The same error will occur。

--- a/entc/gen/type.go
+++ b/entc/gen/type.go
@@ -1203,36 +1203,55 @@ func structTag(name, tag string) string {
 // and ensures it doesn't conflict with Go keywords and other
 // builder fields and it's not exported.
 func builderField(name string) string {
-	if token.Lookup(name).IsKeyword() || name == "config" || strings.ToUpper(name[:1]) == name[:1] {
+	_, ok := privateField[name]
+	if ok || token.Lookup(name).IsKeyword() || strings.ToUpper(name[:1]) == name[:1] {
 		return "_" + name
 	}
 	return name
 }
 
-// global identifiers used by the generated package.
-var globalIdent = names(
-	"AggregateFunc",
-	"As",
-	"Asc",
-	"Count",
-	"Debug",
-	"Desc",
-	"Driver",
-	"Hook",
-	"Log",
-	"MutateFunc",
-	"Mutation",
-	"Mutator",
-	"Op",
-	"Option",
-	"OrderFunc",
-	"Max",
-	"Mean",
-	"Min",
-	"Sum",
-	"Policy",
-	"Query",
-	"Value",
+var (
+	// global identifiers used by the generated package.
+	globalIdent = names(
+		"AggregateFunc",
+		"As",
+		"Asc",
+		"Count",
+		"Debug",
+		"Desc",
+		"Driver",
+		"Hook",
+		"Log",
+		"MutateFunc",
+		"Mutation",
+		"Mutator",
+		"Op",
+		"Option",
+		"OrderFunc",
+		"Max",
+		"Mean",
+		"Min",
+		"Sum",
+		"Policy",
+		"Query",
+		"Value",
+	)
+	// private fields used by the different builders.
+	privateField = names(
+		"config",
+		"done",
+		"hooks",
+		"limit",
+		"mutation",
+		"offset",
+		"oldValue",
+		"order",
+		"op",
+		"path",
+		"predicates",
+		"typ",
+		"unique",
+	)
 )
 
 func names(ids ...string) map[string]struct{} {

--- a/entc/integration/ent/file.go
+++ b/entc/integration/ent/file.go
@@ -29,6 +29,8 @@ type File struct {
 	User *string `json:"user,omitempty"`
 	// Group holds the value of the "group" field.
 	Group string `json:"group,omitempty"`
+	// Op holds the value of the "op" field.
+	Op bool `json:"op,omitempty"`
 	// Edges holds the relations/edges for other nodes in the graph.
 	// The values are being populated by the FileQuery when eager-loading is set.
 	Edges           FileEdges `json:"edges"`
@@ -95,6 +97,7 @@ func (*File) scanValues() []interface{} {
 		&sql.NullString{}, // name
 		&sql.NullString{}, // user
 		&sql.NullString{}, // group
+		&sql.NullBool{},   // op
 	}
 }
 
@@ -140,7 +143,12 @@ func (f *File) assignValues(values ...interface{}) error {
 	} else if value.Valid {
 		f.Group = value.String
 	}
-	values = values[4:]
+	if value, ok := values[4].(*sql.NullBool); !ok {
+		return fmt.Errorf("unexpected type %T for field op", values[4])
+	} else if value.Valid {
+		f.Op = value.Bool
+	}
+	values = values[5:]
 	if len(values) == len(file.ForeignKeys) {
 		if value, ok := values[0].(*sql.NullInt64); !ok {
 			return fmt.Errorf("unexpected type %T for edge-field file_type_files", value)
@@ -212,6 +220,8 @@ func (f *File) String() string {
 	}
 	builder.WriteString(", group=")
 	builder.WriteString(f.Group)
+	builder.WriteString(", op=")
+	builder.WriteString(fmt.Sprintf("%v", f.Op))
 	builder.WriteByte(')')
 	return builder.String()
 }

--- a/entc/integration/ent/file/file.go
+++ b/entc/integration/ent/file/file.go
@@ -19,6 +19,8 @@ const (
 	FieldUser = "user"
 	// FieldGroup holds the string denoting the group field in the database.
 	FieldGroup = "group"
+	// FieldOp holds the string denoting the op field in the database.
+	FieldOp = "op"
 
 	// EdgeOwner holds the string denoting the owner edge name in mutations.
 	EdgeOwner = "owner"
@@ -59,6 +61,7 @@ var Columns = []string{
 	FieldName,
 	FieldUser,
 	FieldGroup,
+	FieldOp,
 }
 
 // ForeignKeys holds the SQL foreign-keys that are owned by the File type.

--- a/entc/integration/ent/file/where.go
+++ b/entc/integration/ent/file/where.go
@@ -123,6 +123,13 @@ func Group(v string) predicate.File {
 	})
 }
 
+// Op applies equality check predicate on the "op" field. It's identical to OpEQ.
+func Op(v bool) predicate.File {
+	return predicate.File(func(s *sql.Selector) {
+		s.Where(sql.EQ(s.C(FieldOp), v))
+	})
+}
+
 // SizeEQ applies the EQ predicate on the "size" field.
 func SizeEQ(v int) predicate.File {
 	return predicate.File(func(s *sql.Selector) {
@@ -557,6 +564,34 @@ func GroupEqualFold(v string) predicate.File {
 func GroupContainsFold(v string) predicate.File {
 	return predicate.File(func(s *sql.Selector) {
 		s.Where(sql.ContainsFold(s.C(FieldGroup), v))
+	})
+}
+
+// OpEQ applies the EQ predicate on the "op" field.
+func OpEQ(v bool) predicate.File {
+	return predicate.File(func(s *sql.Selector) {
+		s.Where(sql.EQ(s.C(FieldOp), v))
+	})
+}
+
+// OpNEQ applies the NEQ predicate on the "op" field.
+func OpNEQ(v bool) predicate.File {
+	return predicate.File(func(s *sql.Selector) {
+		s.Where(sql.NEQ(s.C(FieldOp), v))
+	})
+}
+
+// OpIsNil applies the IsNil predicate on the "op" field.
+func OpIsNil() predicate.File {
+	return predicate.File(func(s *sql.Selector) {
+		s.Where(sql.IsNull(s.C(FieldOp)))
+	})
+}
+
+// OpNotNil applies the NotNil predicate on the "op" field.
+func OpNotNil() predicate.File {
+	return predicate.File(func(s *sql.Selector) {
+		s.Where(sql.NotNull(s.C(FieldOp)))
 	})
 }
 

--- a/entc/integration/ent/file_create.go
+++ b/entc/integration/ent/file_create.go
@@ -74,6 +74,20 @@ func (fc *FileCreate) SetNillableGroup(s *string) *FileCreate {
 	return fc
 }
 
+// SetOp sets the op field.
+func (fc *FileCreate) SetOp(b bool) *FileCreate {
+	fc.mutation.SetOp(b)
+	return fc
+}
+
+// SetNillableOp sets the op field if the given value is not nil.
+func (fc *FileCreate) SetNillableOp(b *bool) *FileCreate {
+	if b != nil {
+		fc.SetOp(*b)
+	}
+	return fc
+}
+
 // SetOwnerID sets the owner edge to User by id.
 func (fc *FileCreate) SetOwnerID(id int) *FileCreate {
 	fc.mutation.SetOwnerID(id)
@@ -256,6 +270,14 @@ func (fc *FileCreate) createSpec() (*File, *sqlgraph.CreateSpec) {
 			Column: file.FieldGroup,
 		})
 		f.Group = value
+	}
+	if value, ok := fc.mutation.GetOp(); ok {
+		_spec.Fields = append(_spec.Fields, &sqlgraph.FieldSpec{
+			Type:   field.TypeBool,
+			Value:  value,
+			Column: file.FieldOp,
+		})
+		f.Op = value
 	}
 	if nodes := fc.mutation.OwnerIDs(); len(nodes) > 0 {
 		edge := &sqlgraph.EdgeSpec{

--- a/entc/integration/ent/file_update.go
+++ b/entc/integration/ent/file_update.go
@@ -101,6 +101,26 @@ func (fu *FileUpdate) ClearGroup() *FileUpdate {
 	return fu
 }
 
+// SetOp sets the op field.
+func (fu *FileUpdate) SetOp(b bool) *FileUpdate {
+	fu.mutation.SetOp(b)
+	return fu
+}
+
+// SetNillableOp sets the op field if the given value is not nil.
+func (fu *FileUpdate) SetNillableOp(b *bool) *FileUpdate {
+	if b != nil {
+		fu.SetOp(*b)
+	}
+	return fu
+}
+
+// ClearOp clears the value of op.
+func (fu *FileUpdate) ClearOp() *FileUpdate {
+	fu.mutation.ClearOp()
+	return fu
+}
+
 // SetOwnerID sets the owner edge to User by id.
 func (fu *FileUpdate) SetOwnerID(id int) *FileUpdate {
 	fu.mutation.SetOwnerID(id)
@@ -324,6 +344,19 @@ func (fu *FileUpdate) sqlSave(ctx context.Context) (n int, err error) {
 			Column: file.FieldGroup,
 		})
 	}
+	if value, ok := fu.mutation.GetOp(); ok {
+		_spec.Fields.Set = append(_spec.Fields.Set, &sqlgraph.FieldSpec{
+			Type:   field.TypeBool,
+			Value:  value,
+			Column: file.FieldOp,
+		})
+	}
+	if fu.mutation.OpCleared() {
+		_spec.Fields.Clear = append(_spec.Fields.Clear, &sqlgraph.FieldSpec{
+			Type:   field.TypeBool,
+			Column: file.FieldOp,
+		})
+	}
 	if fu.mutation.OwnerCleared() {
 		edge := &sqlgraph.EdgeSpec{
 			Rel:     sqlgraph.M2O,
@@ -530,6 +563,26 @@ func (fuo *FileUpdateOne) SetNillableGroup(s *string) *FileUpdateOne {
 // ClearGroup clears the value of group.
 func (fuo *FileUpdateOne) ClearGroup() *FileUpdateOne {
 	fuo.mutation.ClearGroup()
+	return fuo
+}
+
+// SetOp sets the op field.
+func (fuo *FileUpdateOne) SetOp(b bool) *FileUpdateOne {
+	fuo.mutation.SetOp(b)
+	return fuo
+}
+
+// SetNillableOp sets the op field if the given value is not nil.
+func (fuo *FileUpdateOne) SetNillableOp(b *bool) *FileUpdateOne {
+	if b != nil {
+		fuo.SetOp(*b)
+	}
+	return fuo
+}
+
+// ClearOp clears the value of op.
+func (fuo *FileUpdateOne) ClearOp() *FileUpdateOne {
+	fuo.mutation.ClearOp()
 	return fuo
 }
 
@@ -752,6 +805,19 @@ func (fuo *FileUpdateOne) sqlSave(ctx context.Context) (f *File, err error) {
 		_spec.Fields.Clear = append(_spec.Fields.Clear, &sqlgraph.FieldSpec{
 			Type:   field.TypeString,
 			Column: file.FieldGroup,
+		})
+	}
+	if value, ok := fuo.mutation.GetOp(); ok {
+		_spec.Fields.Set = append(_spec.Fields.Set, &sqlgraph.FieldSpec{
+			Type:   field.TypeBool,
+			Value:  value,
+			Column: file.FieldOp,
+		})
+	}
+	if fuo.mutation.OpCleared() {
+		_spec.Fields.Clear = append(_spec.Fields.Clear, &sqlgraph.FieldSpec{
+			Type:   field.TypeBool,
+			Column: file.FieldOp,
 		})
 	}
 	if fuo.mutation.OwnerCleared() {

--- a/entc/integration/ent/migrate/schema.go
+++ b/entc/integration/ent/migrate/schema.go
@@ -122,6 +122,7 @@ var (
 		{Name: "name", Type: field.TypeString},
 		{Name: "user", Type: field.TypeString, Nullable: true},
 		{Name: "group", Type: field.TypeString, Nullable: true},
+		{Name: "op", Type: field.TypeBool, Nullable: true},
 		{Name: "file_type_files", Type: field.TypeInt, Nullable: true},
 		{Name: "group_files", Type: field.TypeInt, Nullable: true},
 		{Name: "user_files", Type: field.TypeInt, Nullable: true},
@@ -134,21 +135,21 @@ var (
 		ForeignKeys: []*schema.ForeignKey{
 			{
 				Symbol:  "files_file_types_files",
-				Columns: []*schema.Column{FilesColumns[5]},
+				Columns: []*schema.Column{FilesColumns[6]},
 
 				RefColumns: []*schema.Column{FileTypesColumns[0]},
 				OnDelete:   schema.SetNull,
 			},
 			{
 				Symbol:  "files_groups_files",
-				Columns: []*schema.Column{FilesColumns[6]},
+				Columns: []*schema.Column{FilesColumns[7]},
 
 				RefColumns: []*schema.Column{GroupsColumns[0]},
 				OnDelete:   schema.SetNull,
 			},
 			{
 				Symbol:  "files_users_files",
-				Columns: []*schema.Column{FilesColumns[7]},
+				Columns: []*schema.Column{FilesColumns[8]},
 
 				RefColumns: []*schema.Column{UsersColumns[0]},
 				OnDelete:   schema.SetNull,
@@ -168,17 +169,17 @@ var (
 			{
 				Name:    "file_user_files_file_type_files",
 				Unique:  false,
-				Columns: []*schema.Column{FilesColumns[7], FilesColumns[5]},
+				Columns: []*schema.Column{FilesColumns[8], FilesColumns[6]},
 			},
 			{
 				Name:    "file_name_user_files_file_type_files",
 				Unique:  true,
-				Columns: []*schema.Column{FilesColumns[2], FilesColumns[7], FilesColumns[5]},
+				Columns: []*schema.Column{FilesColumns[2], FilesColumns[8], FilesColumns[6]},
 			},
 			{
 				Name:    "file_name_user_files",
 				Unique:  false,
-				Columns: []*schema.Column{FilesColumns[2], FilesColumns[7]},
+				Columns: []*schema.Column{FilesColumns[2], FilesColumns[8]},
 			},
 		},
 	}

--- a/entc/integration/ent/mutation.go
+++ b/entc/integration/ent/mutation.go
@@ -5669,6 +5669,7 @@ type FileMutation struct {
 	name          *string
 	user          *string
 	group         *string
+	_op           *bool
 	clearedFields map[string]struct{}
 	owner         *int
 	clearedowner  bool
@@ -5954,6 +5955,56 @@ func (m *FileMutation) ResetGroup() {
 	delete(m.clearedFields, file.FieldGroup)
 }
 
+// SetOp sets the op field.
+func (m *FileMutation) SetOp(b bool) {
+	m._op = &b
+}
+
+// GetOp returns the op value in the mutation.
+func (m *FileMutation) GetOp() (r bool, exists bool) {
+	v := m._op
+	if v == nil {
+		return
+	}
+	return *v, true
+}
+
+// OldOp returns the old op value of the File.
+// If the File object wasn't provided to the builder, the object is fetched
+// from the database.
+// An error is returned if the mutation operation is not UpdateOne, or database query fails.
+func (m *FileMutation) OldOp(ctx context.Context) (v bool, err error) {
+	if !m.op.Is(OpUpdateOne) {
+		return v, fmt.Errorf("OldOp is allowed only on UpdateOne operations")
+	}
+	if m.id == nil || m.oldValue == nil {
+		return v, fmt.Errorf("OldOp requires an ID field in the mutation")
+	}
+	oldValue, err := m.oldValue(ctx)
+	if err != nil {
+		return v, fmt.Errorf("querying old value for OldOp: %w", err)
+	}
+	return oldValue.Op, nil
+}
+
+// ClearOp clears the value of op.
+func (m *FileMutation) ClearOp() {
+	m._op = nil
+	m.clearedFields[file.FieldOp] = struct{}{}
+}
+
+// OpCleared returns if the field op was cleared in this mutation.
+func (m *FileMutation) OpCleared() bool {
+	_, ok := m.clearedFields[file.FieldOp]
+	return ok
+}
+
+// ResetOp reset all changes of the "op" field.
+func (m *FileMutation) ResetOp() {
+	m._op = nil
+	delete(m.clearedFields, file.FieldOp)
+}
+
 // SetOwnerID sets the owner edge to User by id.
 func (m *FileMutation) SetOwnerID(id int) {
 	m.owner = &id
@@ -6099,7 +6150,7 @@ func (m *FileMutation) Type() string {
 // this mutation. Note that, in order to get all numeric
 // fields that were in/decremented, call AddedFields().
 func (m *FileMutation) Fields() []string {
-	fields := make([]string, 0, 4)
+	fields := make([]string, 0, 5)
 	if m.size != nil {
 		fields = append(fields, file.FieldSize)
 	}
@@ -6111,6 +6162,9 @@ func (m *FileMutation) Fields() []string {
 	}
 	if m.group != nil {
 		fields = append(fields, file.FieldGroup)
+	}
+	if m._op != nil {
+		fields = append(fields, file.FieldOp)
 	}
 	return fields
 }
@@ -6128,6 +6182,8 @@ func (m *FileMutation) Field(name string) (ent.Value, bool) {
 		return m.User()
 	case file.FieldGroup:
 		return m.Group()
+	case file.FieldOp:
+		return m.GetOp()
 	}
 	return nil, false
 }
@@ -6145,6 +6201,8 @@ func (m *FileMutation) OldField(ctx context.Context, name string) (ent.Value, er
 		return m.OldUser(ctx)
 	case file.FieldGroup:
 		return m.OldGroup(ctx)
+	case file.FieldOp:
+		return m.OldOp(ctx)
 	}
 	return nil, fmt.Errorf("unknown File field %s", name)
 }
@@ -6181,6 +6239,13 @@ func (m *FileMutation) SetField(name string, value ent.Value) error {
 			return fmt.Errorf("unexpected type %T for field %s", value, name)
 		}
 		m.SetGroup(v)
+		return nil
+	case file.FieldOp:
+		v, ok := value.(bool)
+		if !ok {
+			return fmt.Errorf("unexpected type %T for field %s", value, name)
+		}
+		m.SetOp(v)
 		return nil
 	}
 	return fmt.Errorf("unknown File field %s", name)
@@ -6233,6 +6298,9 @@ func (m *FileMutation) ClearedFields() []string {
 	if m.FieldCleared(file.FieldGroup) {
 		fields = append(fields, file.FieldGroup)
 	}
+	if m.FieldCleared(file.FieldOp) {
+		fields = append(fields, file.FieldOp)
+	}
 	return fields
 }
 
@@ -6252,6 +6320,9 @@ func (m *FileMutation) ClearField(name string) error {
 		return nil
 	case file.FieldGroup:
 		m.ClearGroup()
+		return nil
+	case file.FieldOp:
+		m.ClearOp()
 		return nil
 	}
 	return fmt.Errorf("unknown File nullable field %s", name)
@@ -6273,6 +6344,9 @@ func (m *FileMutation) ResetField(name string) error {
 		return nil
 	case file.FieldGroup:
 		m.ResetGroup()
+		return nil
+	case file.FieldOp:
+		m.ResetOp()
 		return nil
 	}
 	return fmt.Errorf("unknown File field %s", name)

--- a/entc/integration/ent/schema/file.go
+++ b/entc/integration/ent/schema/file.go
@@ -31,6 +31,8 @@ func (File) Fields() []ent.Field {
 			Nillable(),
 		field.String("group").
 			Optional(),
+		field.Bool("op").
+			Optional(),
 	}
 }
 

--- a/entc/integration/gremlin/ent/file.go
+++ b/entc/integration/gremlin/ent/file.go
@@ -28,6 +28,8 @@ type File struct {
 	User *string `json:"user,omitempty"`
 	// Group holds the value of the "group" field.
 	Group string `json:"group,omitempty"`
+	// Op holds the value of the "op" field.
+	Op bool `json:"op,omitempty"`
 	// Edges holds the relations/edges for other nodes in the graph.
 	// The values are being populated by the FileQuery when eager-loading is set.
 	Edges FileEdges `json:"edges"`
@@ -95,6 +97,7 @@ func (f *File) FromResponse(res *gremlin.Response) error {
 		Name  string  `json:"name,omitempty"`
 		User  *string `json:"user,omitempty"`
 		Group string  `json:"group,omitempty"`
+		Op    bool    `json:"op,omitempty"`
 	}
 	if err := vmap.Decode(&scanf); err != nil {
 		return err
@@ -104,6 +107,7 @@ func (f *File) FromResponse(res *gremlin.Response) error {
 	f.Name = scanf.Name
 	f.User = scanf.User
 	f.Group = scanf.Group
+	f.Op = scanf.Op
 	return nil
 }
 
@@ -155,6 +159,8 @@ func (f *File) String() string {
 	}
 	builder.WriteString(", group=")
 	builder.WriteString(f.Group)
+	builder.WriteString(", op=")
+	builder.WriteString(fmt.Sprintf("%v", f.Op))
 	builder.WriteByte(')')
 	return builder.String()
 }
@@ -174,6 +180,7 @@ func (f *Files) FromResponse(res *gremlin.Response) error {
 		Name  string  `json:"name,omitempty"`
 		User  *string `json:"user,omitempty"`
 		Group string  `json:"group,omitempty"`
+		Op    bool    `json:"op,omitempty"`
 	}
 	if err := vmap.Decode(&scanf); err != nil {
 		return err
@@ -185,6 +192,7 @@ func (f *Files) FromResponse(res *gremlin.Response) error {
 			Name:  v.Name,
 			User:  v.User,
 			Group: v.Group,
+			Op:    v.Op,
 		})
 	}
 	return nil

--- a/entc/integration/gremlin/ent/file/file.go
+++ b/entc/integration/gremlin/ent/file/file.go
@@ -19,6 +19,8 @@ const (
 	FieldUser = "user"
 	// FieldGroup holds the string denoting the group field in the database.
 	FieldGroup = "group"
+	// FieldOp holds the string denoting the op field in the database.
+	FieldOp = "op"
 
 	// EdgeOwner holds the string denoting the owner edge name in mutations.
 	EdgeOwner = "owner"

--- a/entc/integration/gremlin/ent/file/where.go
+++ b/entc/integration/gremlin/ent/file/where.go
@@ -112,6 +112,13 @@ func Group(v string) predicate.File {
 	})
 }
 
+// Op applies equality check predicate on the "op" field. It's identical to OpEQ.
+func Op(v bool) predicate.File {
+	return predicate.File(func(t *dsl.Traversal) {
+		t.Has(Label, FieldOp, p.EQ(v))
+	})
+}
+
 // SizeEQ applies the EQ predicate on the "size" field.
 func SizeEQ(v int) predicate.File {
 	return predicate.File(func(t *dsl.Traversal) {
@@ -456,6 +463,34 @@ func GroupIsNil() predicate.File {
 func GroupNotNil() predicate.File {
 	return predicate.File(func(t *dsl.Traversal) {
 		t.HasLabel(Label).Has(FieldGroup)
+	})
+}
+
+// OpEQ applies the EQ predicate on the "op" field.
+func OpEQ(v bool) predicate.File {
+	return predicate.File(func(t *dsl.Traversal) {
+		t.Has(Label, FieldOp, p.EQ(v))
+	})
+}
+
+// OpNEQ applies the NEQ predicate on the "op" field.
+func OpNEQ(v bool) predicate.File {
+	return predicate.File(func(t *dsl.Traversal) {
+		t.Has(Label, FieldOp, p.NEQ(v))
+	})
+}
+
+// OpIsNil applies the IsNil predicate on the "op" field.
+func OpIsNil() predicate.File {
+	return predicate.File(func(t *dsl.Traversal) {
+		t.HasLabel(Label).HasNot(FieldOp)
+	})
+}
+
+// OpNotNil applies the NotNil predicate on the "op" field.
+func OpNotNil() predicate.File {
+	return predicate.File(func(t *dsl.Traversal) {
+		t.HasLabel(Label).Has(FieldOp)
 	})
 }
 

--- a/entc/integration/gremlin/ent/file_create.go
+++ b/entc/integration/gremlin/ent/file_create.go
@@ -76,6 +76,20 @@ func (fc *FileCreate) SetNillableGroup(s *string) *FileCreate {
 	return fc
 }
 
+// SetOp sets the op field.
+func (fc *FileCreate) SetOp(b bool) *FileCreate {
+	fc.mutation.SetOp(b)
+	return fc
+}
+
+// SetNillableOp sets the op field if the given value is not nil.
+func (fc *FileCreate) SetNillableOp(b *bool) *FileCreate {
+	if b != nil {
+		fc.SetOp(*b)
+	}
+	return fc
+}
+
 // SetOwnerID sets the owner edge to User by id.
 func (fc *FileCreate) SetOwnerID(id string) *FileCreate {
 	fc.mutation.SetOwnerID(id)
@@ -237,6 +251,9 @@ func (fc *FileCreate) gremlin() *dsl.Traversal {
 	}
 	if value, ok := fc.mutation.Group(); ok {
 		v.Property(dsl.Single, file.FieldGroup, value)
+	}
+	if value, ok := fc.mutation.GetOp(); ok {
+		v.Property(dsl.Single, file.FieldOp, value)
 	}
 	for _, id := range fc.mutation.OwnerIDs() {
 		v.AddE(user.FilesLabel).From(g.V(id)).InV()

--- a/entc/integration/gremlin/ent/file_update.go
+++ b/entc/integration/gremlin/ent/file_update.go
@@ -102,6 +102,26 @@ func (fu *FileUpdate) ClearGroup() *FileUpdate {
 	return fu
 }
 
+// SetOp sets the op field.
+func (fu *FileUpdate) SetOp(b bool) *FileUpdate {
+	fu.mutation.SetOp(b)
+	return fu
+}
+
+// SetNillableOp sets the op field if the given value is not nil.
+func (fu *FileUpdate) SetNillableOp(b *bool) *FileUpdate {
+	if b != nil {
+		fu.SetOp(*b)
+	}
+	return fu
+}
+
+// ClearOp clears the value of op.
+func (fu *FileUpdate) ClearOp() *FileUpdate {
+	fu.mutation.ClearOp()
+	return fu
+}
+
 // SetOwnerID sets the owner edge to User by id.
 func (fu *FileUpdate) SetOwnerID(id string) *FileUpdate {
 	fu.mutation.SetOwnerID(id)
@@ -303,12 +323,18 @@ func (fu *FileUpdate) gremlin() *dsl.Traversal {
 	if value, ok := fu.mutation.Group(); ok {
 		v.Property(dsl.Single, file.FieldGroup, value)
 	}
+	if value, ok := fu.mutation.GetOp(); ok {
+		v.Property(dsl.Single, file.FieldOp, value)
+	}
 	var properties []interface{}
 	if fu.mutation.UserCleared() {
 		properties = append(properties, file.FieldUser)
 	}
 	if fu.mutation.GroupCleared() {
 		properties = append(properties, file.FieldGroup)
+	}
+	if fu.mutation.OpCleared() {
+		properties = append(properties, file.FieldOp)
 	}
 	if len(properties) > 0 {
 		v.SideEffect(__.Properties(properties...).Drop())
@@ -424,6 +450,26 @@ func (fuo *FileUpdateOne) SetNillableGroup(s *string) *FileUpdateOne {
 // ClearGroup clears the value of group.
 func (fuo *FileUpdateOne) ClearGroup() *FileUpdateOne {
 	fuo.mutation.ClearGroup()
+	return fuo
+}
+
+// SetOp sets the op field.
+func (fuo *FileUpdateOne) SetOp(b bool) *FileUpdateOne {
+	fuo.mutation.SetOp(b)
+	return fuo
+}
+
+// SetNillableOp sets the op field if the given value is not nil.
+func (fuo *FileUpdateOne) SetNillableOp(b *bool) *FileUpdateOne {
+	if b != nil {
+		fuo.SetOp(*b)
+	}
+	return fuo
+}
+
+// ClearOp clears the value of op.
+func (fuo *FileUpdateOne) ClearOp() *FileUpdateOne {
+	fuo.mutation.ClearOp()
 	return fuo
 }
 
@@ -633,12 +679,18 @@ func (fuo *FileUpdateOne) gremlin(id string) *dsl.Traversal {
 	if value, ok := fuo.mutation.Group(); ok {
 		v.Property(dsl.Single, file.FieldGroup, value)
 	}
+	if value, ok := fuo.mutation.GetOp(); ok {
+		v.Property(dsl.Single, file.FieldOp, value)
+	}
 	var properties []interface{}
 	if fuo.mutation.UserCleared() {
 		properties = append(properties, file.FieldUser)
 	}
 	if fuo.mutation.GroupCleared() {
 		properties = append(properties, file.FieldGroup)
+	}
+	if fuo.mutation.OpCleared() {
+		properties = append(properties, file.FieldOp)
 	}
 	if len(properties) > 0 {
 		v.SideEffect(__.Properties(properties...).Drop())

--- a/entc/integration/gremlin/ent/mutation.go
+++ b/entc/integration/gremlin/ent/mutation.go
@@ -5669,6 +5669,7 @@ type FileMutation struct {
 	name          *string
 	user          *string
 	group         *string
+	_op           *bool
 	clearedFields map[string]struct{}
 	owner         *string
 	clearedowner  bool
@@ -5954,6 +5955,56 @@ func (m *FileMutation) ResetGroup() {
 	delete(m.clearedFields, file.FieldGroup)
 }
 
+// SetOp sets the op field.
+func (m *FileMutation) SetOp(b bool) {
+	m._op = &b
+}
+
+// GetOp returns the op value in the mutation.
+func (m *FileMutation) GetOp() (r bool, exists bool) {
+	v := m._op
+	if v == nil {
+		return
+	}
+	return *v, true
+}
+
+// OldOp returns the old op value of the File.
+// If the File object wasn't provided to the builder, the object is fetched
+// from the database.
+// An error is returned if the mutation operation is not UpdateOne, or database query fails.
+func (m *FileMutation) OldOp(ctx context.Context) (v bool, err error) {
+	if !m.op.Is(OpUpdateOne) {
+		return v, fmt.Errorf("OldOp is allowed only on UpdateOne operations")
+	}
+	if m.id == nil || m.oldValue == nil {
+		return v, fmt.Errorf("OldOp requires an ID field in the mutation")
+	}
+	oldValue, err := m.oldValue(ctx)
+	if err != nil {
+		return v, fmt.Errorf("querying old value for OldOp: %w", err)
+	}
+	return oldValue.Op, nil
+}
+
+// ClearOp clears the value of op.
+func (m *FileMutation) ClearOp() {
+	m._op = nil
+	m.clearedFields[file.FieldOp] = struct{}{}
+}
+
+// OpCleared returns if the field op was cleared in this mutation.
+func (m *FileMutation) OpCleared() bool {
+	_, ok := m.clearedFields[file.FieldOp]
+	return ok
+}
+
+// ResetOp reset all changes of the "op" field.
+func (m *FileMutation) ResetOp() {
+	m._op = nil
+	delete(m.clearedFields, file.FieldOp)
+}
+
 // SetOwnerID sets the owner edge to User by id.
 func (m *FileMutation) SetOwnerID(id string) {
 	m.owner = &id
@@ -6099,7 +6150,7 @@ func (m *FileMutation) Type() string {
 // this mutation. Note that, in order to get all numeric
 // fields that were in/decremented, call AddedFields().
 func (m *FileMutation) Fields() []string {
-	fields := make([]string, 0, 4)
+	fields := make([]string, 0, 5)
 	if m.size != nil {
 		fields = append(fields, file.FieldSize)
 	}
@@ -6111,6 +6162,9 @@ func (m *FileMutation) Fields() []string {
 	}
 	if m.group != nil {
 		fields = append(fields, file.FieldGroup)
+	}
+	if m._op != nil {
+		fields = append(fields, file.FieldOp)
 	}
 	return fields
 }
@@ -6128,6 +6182,8 @@ func (m *FileMutation) Field(name string) (ent.Value, bool) {
 		return m.User()
 	case file.FieldGroup:
 		return m.Group()
+	case file.FieldOp:
+		return m.GetOp()
 	}
 	return nil, false
 }
@@ -6145,6 +6201,8 @@ func (m *FileMutation) OldField(ctx context.Context, name string) (ent.Value, er
 		return m.OldUser(ctx)
 	case file.FieldGroup:
 		return m.OldGroup(ctx)
+	case file.FieldOp:
+		return m.OldOp(ctx)
 	}
 	return nil, fmt.Errorf("unknown File field %s", name)
 }
@@ -6181,6 +6239,13 @@ func (m *FileMutation) SetField(name string, value ent.Value) error {
 			return fmt.Errorf("unexpected type %T for field %s", value, name)
 		}
 		m.SetGroup(v)
+		return nil
+	case file.FieldOp:
+		v, ok := value.(bool)
+		if !ok {
+			return fmt.Errorf("unexpected type %T for field %s", value, name)
+		}
+		m.SetOp(v)
 		return nil
 	}
 	return fmt.Errorf("unknown File field %s", name)
@@ -6233,6 +6298,9 @@ func (m *FileMutation) ClearedFields() []string {
 	if m.FieldCleared(file.FieldGroup) {
 		fields = append(fields, file.FieldGroup)
 	}
+	if m.FieldCleared(file.FieldOp) {
+		fields = append(fields, file.FieldOp)
+	}
 	return fields
 }
 
@@ -6252,6 +6320,9 @@ func (m *FileMutation) ClearField(name string) error {
 		return nil
 	case file.FieldGroup:
 		m.ClearGroup()
+		return nil
+	case file.FieldOp:
+		m.ClearOp()
 		return nil
 	}
 	return fmt.Errorf("unknown File nullable field %s", name)
@@ -6273,6 +6344,9 @@ func (m *FileMutation) ResetField(name string) error {
 		return nil
 	case file.FieldGroup:
 		m.ResetGroup()
+		return nil
+	case file.FieldOp:
+		m.ResetOp()
 		return nil
 	}
 	return fmt.Errorf("unknown File field %s", name)

